### DISCUSSION TITLE
how to include both darwin and linux in ortools-java jar?

### DISCUSSION THREAD

**Discussion by ben-xj** [google/or-tools#3733] [Build and configuration issues]
**What version of OR-Tools and what language are you using?**
Version: v9.6
Language: Java

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
SCIP

**What operating system (Linux, Windows, ...) and version?**
Macos monterey & centos 7

**What did you do?**
1. build the jar pack from source on mac 
2. works well in my project on my mac
3. maven packaged the project to a jar and deployed the jar on centos

**What did you expect to see**
works well as on my mac

**What did you see instead?**
```
java.lang.NullPointerException: Resource "ortools-linux-x86-64/libortools.so" was not found in location "com.google.ortools.HasResources$$Lambda$552/1352604812@63790529"
        at java.util.Objects.requireNonNull(Objects.java:228)
        at com.google.ortools.Loader.extractLibrary(Loader.java:93)
        at com.google.ortools.Loader.loadNativeLibraries(Loader.java:69)
        at com.google.ortools.Loader.loadNativeLibraries(Loader.java:81)
```

Make sure you include information that can help us debug (full error message, model Proto).
I found that only ortools-darwin-x86-64 is included in the jar, not   ortools-linux-x86-64, how can i include  ortools-darwin-x86-64 as well?

**Anything else we should know about your project / environment**

**Comment by Mizux** [comment]
two points:
at configure/compile time we are using this CMake function to determine the current host
https://github.com/google/or-tools/blob/7a0fd2ceb8e1d3bda29c3e77823e13aea4355071/cmake/java.cmake#L59-L76
note: since we introspect the host to build the maven package you need a linux env to build the linux package etc...

Then at runtime we use [`jna`](https://mvnrepository.com/artifact/net.java.dev.jna/jna-platform) to retrieve the current platform and load the maven package accordingly

see:
https://github.com/google/or-tools/blob/7a0fd2ceb8e1d3bda29c3e77823e13aea4355071/ortools/java/com/google/ortools/Loader.java#L34

So to deploy for centos you need to use at least a linux docker, for example you may use the ones we use to build artifacts in tools/docker
```sh
cd tools/docker
make # to get help usage
make amd64_centos-7_java_export # should build a centos artifact
```
note: not sure it will work with your fork you may need to tweak the Dockerfile
https://github.com/google/or-tools/blob/7a0fd2ceb8e1d3bda29c3e77823e13aea4355071/tools/docker/images/centos-7.Dockerfile#L93-L94
please notice your changes must be submitted in your current branch since the Dockerfile will perform a `git clone ...` to retrieve the source.

otherwise you also have the ci jobs in cmake/ but they won't export the maven artifact you need to do it manually
```sh
cd cmake
make # see usage
make centos_java_build # to build an image with the java package
make sh_centos_java_build # to spawn a container using it to inspect...
```
Dockerfile:
https://github.com/google/or-tools/blob/main/cmake/docker/centos/Dockerfile
https://github.com/google/or-tools/blob/main/cmake/docker/centos/java.Dockerfile

tips: to get more trace you can use `export BUILDKIT_PROGRESS=plain` nowadays docker hide them by default `-_-`
